academic_observatory_workflows/workflows/openalex_telescope.py

```diff
@@ -356,53 +363,70 @@ def __init__(
         self.aws_openalex_bucket = aws_openalex_bucket
         self.observatory_api_conn_id = observatory_api_conn_id
 
-        # Wait for the previous DAG run to finish to make sure that
-        # changefiles are processed in the correct order
-        external_task_id = "dag_run_complete"
-        self.add_operator(
-            PreviousDagRunSensor(
+    def make_dag(self) -> DAG:
+        with self.dag:
+            # fmt: off
+            # Wait for the previous DAG run to finish to make sure that
+            # changefiles are processed in the correct order
+            external_task_id = "dag_run_complete"
+            task_sensor = PreviousDagRunSensor(
                 dag_id=self.dag_id,
                 external_task_id=external_task_id,
                 execution_delta=timedelta(days=7),  # To match the @weekly schedule
             )
-        )
-        self.add_setup_task(self.check_dependencies)
-        self.add_setup_task(self.fetch_releases)
-        self.add_task(self.create_datasets)
-
-        # Create snapshots of main tables in case we mess up
-        # This is done before updating the tables to make sure that the snapshots haven't expired before the tables
-        # are updated
-        self.add_task(self.bq_create_main_table_snapshots)
-
-        # Transfer, download and transform data
-        self.add_task(self.aws_to_gcs_transfer)
-
-        # Download concepts, institutions and works which need to be pre-processed
-        for entity_name, transform in self.entities:
-            if transform:
-                self.add_operator(make_download_bash_op(self, entity_name))
-        self.add_task(self.transform)
-
-        # Upsert records
-        self.add_task(self.upload_upsert_files)
-        self.add_task(self.bq_load_upsert_tables)
-        self.add_task(self.bq_upsert_records)
-
-        # Delete records
-        self.add_task(self.bq_load_delete_tables)
-        self.add_task(self.bq_delete_records)
-
-        # Add release info to API and cleanup
-        self.add_task(self.add_new_dataset_releases)
-        self.add_task(self.cleanup)
-
-        # The last task that the next DAG run's ExternalTaskSensor waits for.
-        self.add_operator(
-            EmptyOperator(
-                task_id=external_task_id,
+            task_check_dependencies = PythonOperator(python_callable=self.check_dependencies, task_id="check_dependencies")
+            task_fetch_releases = ShortCircuitOperator(python_callable=self.fetch_releases, task_id="fetch_releases") # If there are no changes then skip all below tasks
+            task_create_datasets = PythonOperator(python_callable=self.create_datasets, task_id="create_datasets")
+            task_aws_to_gcs_transfer = self.make_python_operator(self.aws_to_gcs_transfer, "aws_to_gcs_transfer")
+            task_branch = BranchPythonOperator(
```

Comment:
According to the internet, the BranchPythonOperator will push its chosen branch to Xcom by default each time it executes (and not delete the xcom afterwards). This can be changed by adding do_xcom_push=False. Not sure if we want the branch choice flooding the xcom, what do you think?

Reply:
The cleanup task will eventually remove the old XComs, so it should be OK, what do you think?

Reply:
Ah yes this is true, I forgot about the cleanup step :+1: